girder/api/rest.py

```diff
@@ -840,89 +840,90 @@ def handleRoute(self, method, path, params):
+            val = handler(**kwargs)
+
+        # Fire the after-call event that has a chance to augment the
+        # return value of the API method that was called. You can
+        # reassign the return value completely by adding a response to
+        # the event and calling preventDefault() on it.
+        kwargs['returnVal'] = val
+        event = events.trigger('.'.join((eventPrefix, 'after')), kwargs)
+        if event.defaultPrevented and len(event.responses) > 0:
+            val = event.responses[0]
 
-        raise RestException('No matching route for "%s %s"' % (
-            method.upper(), '/'.join(path)))
+        return val
 
-    def _matchRoute(self, path, route):
+    def _matchRoute(self, method, path):
         """
-        Helper function that attempts to match the requested path with a
-        given route specification. Returns False if the requested path does
-        not match the route. If it does match, this will return the dict of
-        kwargs that should be passed to the underlying handler, based on the
-        wildcard tokens of the route.
+        Helper function that attempts to match the requested ``method`` and ``path`` with a
+        registered route specification.
 
+        :param method: The requested HTTP method, in lowercase.
+        :type method: str
         :param path: The requested path.
-        :type path: `list`
-        :param route: The route specification to match against.
-        :type route: `list`
+        :type path: `tuple`
+        :returns: A tuple of ``(route, handler, wildcards)``, where ``route`` is the registered
+                  `list` of route components, ``handler`` is the route handler `function`, and
+                  ``wildcards`` is a `dict` of kwargs that should be passed to the underlying
+                  handler, based on the wildcard tokens of the route.
+        :raises: `GirderException`, when no routes are defined on this resource.
+        :raises: `RestException`, when no route can be matched.
         """
-        wildcards = {}
-        for i in range(0, len(route)):
-            if route[i][0] == ':':  # Wildcard token
-                wildcards[route[i][1:]] = path[i]
-            elif route[i] != path[i]:  # Exact match token
-                return False
-        return wildcards
+        if not self._routes:
+            raise GirderException('No routes defined for resource')
+
+        for route, handler in self._routes[method][len(path)]:
+            wildcards = {}
+            for routeComponent, pathComponent in six.moves.zip(route, path):
+                if routeComponent[0] == ':':  # Wildcard token
+                    wildcards[routeComponent[1:]] = pathComponent
+                elif routeComponent != pathComponent:  # Exact match token
+                    break
+            else:
+                return route, handler, wildcards
+
+        raise RestException('No matching route for "%s %s"' % (method.upper(), '/'.join(path)))
```

Comment:
Are the changes to this file related / necessary to the ToU plugin? If not, could you cherry-pick them onto a separate topic and remove this commit from this one?